src/sbvr-api/permissions.ts

```diff
@@ -1273,6 +1273,96 @@ export const getUserPermissions = async (
 	}
 };
 
+export const getRolePermissions = async (
+	roleName: string,
+	userId: number,
+	tx?: Tx,
+): Promise<string[]> => {
+	try {
+		return await $getRolePermissions(roleName, userId, tx);
+	} catch (err: unknown) {
+		sbvrUtils.api.Auth.logger.error('Error loading role permissions', err);
+		throw err;
+	}
+};
+
+const $getRolePermissions = (() => {
+	const getRolePermissionsQuery = _.once(() =>
+		sbvrUtils.api.Auth.prepare<{ roleName: string; userId: number }>({
+			resource: 'permission',
+			passthrough: {
+				req: rootRead,
+			},
+			options: {
+				$select: 'name',
+				$filter: {
+					is_of__role: {
+						$any: {
+							$alias: 'rhp',
+							$expr: {
+								rhp: {
+									role: {
+										$any: {
+											$alias: 'r',
+											$expr: {
+												r: {
+													name: { '@': 'roleName' },
+													is_of__user: {
+														$any: {
+															$alias: 'uhr',
+															$expr: {
+																uhr: { user: { '@': 'userId' } },
+																$or: [
+																	{
+																		uhr: { expiry_date: null },
+																	},
+																	{
+																		uhr: {
+																			expiry_date: { $gt: { $now: null } },
+																		},
+																	},
+																],
+															},
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				// We orderby to increase the hit rate for the `_checkPermissions` memoisation
+				$orderby: {
+					name: 'asc',
+				},
+			},
+		}),
+	);
+
+	return env.createCache(
+		'rolePermissions',
+		async (roleName: string, userId: number, tx?: Tx) => {
+			const permissions = (await getRolePermissionsQuery()(
+				{
+					roleName,
+					userId,
+				},
+				undefined,
+				{ tx },
+			)) as Array<{ name: string }>;
+			return permissions.map((permission) => permission.name);
+		},
+		{
+			primitive: true,
+			promise: true,
+			normalizer: ([roleName, userId]) => `${roleName}${userId}`,
+		},
+	);
+})();
```

Comment:
I think that the role permissions are independent of the user itself?
If the role permissions can be pulled from the DB without a userId, we should only get them without a user.
Moreover, we should not cache rolePermissions on user+roleName level, but rather on roleName.
Then the user should resolve the cached values and see if the userId has the role and return it.

Reply:
@fisehara sorry - this is not the idea here, the idea is really to grant the role's permission IF the user has this role. This only allows you to generate a token with a smaller scope for all the roles you already have. E.g. if you have 3 roles, instead of creating a token with all the permissions granted by the 3 roles it allows you to create a token with just the permissions of one of these roles - It never gives you more permissions than you already have

Reply:
@fisehara I see. Wouldn't that be worse as it requires two DB roundtrips instead of one?